contracts/Marketplace.sol

```diff
@@ -320,21 +323,31 @@ contract Marketplace is Ownable, IMarketplace2 {
             emit NewSubscription(p.id, subscriber, endTimestamp);
         }
         emit Subscribed(p.id, subscriber, endTimestamp);
-        uint price = 0;
+        
+        uint256 price = 0;
+        uint256 fee = 0;
         if (requirePayment){
             price = getPriceInData(addSeconds, p.pricePerSecond, p.priceCurrency);
-            require(datacoin.transferFrom(msg.sender, p.beneficiary, price), "error_paymentFailed");
+            fee = txFee.mul(price).div(1 ether);
+            require(datacoin.transferFrom(msg.sender, p.beneficiary, price.sub(fee)), "error_paymentFailed");
+            if(fee > 0){
+                require(datacoin.transferFrom(msg.sender, owner, fee), "error_paymentFailed");
+            }
         }
-
-        // Solidity 5:
-        //(bool success, bytes memory returnData) = p.beneficiary.call(abi.encodeWithSignature("onPurchase(bytes32,address,uint256,uint256)", productId, subscriber, oldSub.endTimestamp, price));
-        // TODO: check returnData if onPurchase returned true (accept purchase) or false (reject purchase)
-        // TODO: require(purchaseAccepted, "error_rejectedBySeller")
-
-        // Solidity 4:
-        // 0x91517bdd = keccak256("onPurchase(bytes32,address,uint256,uint256)")
-        // this call returns true if beneficiary is a PurchaseListener, return value is ignored
-        p.beneficiary.call(0x91517bdd, productId, subscriber, oldSub.endTimestamp, price);
+        
+        uint256 codeSize;
+        address addr = p.beneficiary;
+        assembly { codeSize := extcodesize(addr) }  // solhint-disable-line no-inline-assembly
+        if (codeSize > 0) {
+            (bool success, bytes memory returnData) = p.beneficiary.call(abi.encodeWithSignature("onPurchase(bytes32,address,uint256,uint256,uint256)",
```

Comment:
need to check if address is a contract here. I noticed that call() returns success = true when address is a wallet.